### DISCUSSION TITLE
Increase ConfigurationDiscoveryService priority and set initial delay to zero.

### DISCUSSION THREAD

**Discussion by weixiang1862** [apache/skywalking#11246] [Ideas]
I want increase ConfigurationDiscoveryService priority(right after GRPCChannelManager) and  set sync config task initial delay to zero. It can help agent receive config from backend as early as possible.

![image](https://github.com/apache/skywalking/assets/16346644/f41923be-a62e-4b13-a59b-787268f58001)

**Comment by wu-sheng** [comment]
What do you mean by `priority`? I don't think there is such a mechanism in the agent services.

**Reply by weixiang1862** [reply]
`priority` is defined by BootService, it controls the boot service prepare, boot, shutdown order.

**Reply by wu-sheng** [reply]
OK. I can find it now. It was added. But why 20s matter? I can't see your use cases.

**Reply by weixiang1862** [reply]
We publish gray route rule config to configuration center(apollo), it is used by gateway to category http traffic. 
We hope cds can works like bootstrap config, and it can overide default config in config.properties.
So the time is important in this case, or gateway will route incorrectly to upstream service when it boot.

**Reply by wu-sheng** [reply]
That is out of scope. It is not from an APM use case.

As I mentioned in https://github.com/apache/skywalking/discussions/11245#discussioncomment-6778947, it is good for you this works as the fundamental, but don't ask the community for a thing out of its scope.

A project is GOOD because it has a certain and focused landscape. All optimization is the target-oriented trade-off.

**Reply by weixiang1862** [reply]
Yes, I agree, SkyWalking should focus on self. So, let‘s look about a case in skywalking APM use:

User use cds to control agent `agent.ignore_suffix` and `agent.trace.ignore_path`, he may hope this value can overide default value in `agent.config`, although SkyWalking has no such guarantee in [settings override priority](https://skywalking.apache.org/docs/skywalking-java/next/en/setup/service-agent/java-agent/setting-override/#override-priority).

Then, in the first 20s after boot, agent works on default config and later agent works on remote config. I think this inconsistent time can be optimized by cds sync task initial delay.

If you think set initial delay to zero is too extreme, how about make it value configurable in `agent.config` and default to 20s? At least user get the chance to control this.

We already have diff in code with SkyWalking and must do sync between two repo in the feature, we don't mind introduce more changes. I start this discussion just because I think this optimization is meaningful to SkyWalking self. I am a fresh man in SW and has a very narrow view of landscape, I learned a lot from this discussion, thanks. The discuss result is not important, I think and trust you can make the best decision.

![image](https://github.com/apache/skywalking/assets/16346644/1987cebf-4fa8-417a-9b0a-b5eb265d7e3e)

**Reply by wu-sheng** [reply]
Isn't it already in the agent.config?

**Reply by weixiang1862** [reply]
No, just `GET_AGENT_DYNAMIC_CONFIG_INTERVAL`, it controls both initial delay and schedule period.

**Reply by wu-sheng** [reply]
I can't see why the first 20s matters. Most services can't even boot in 20s. 

> Then, in the first 20s after boot, agent works on default config and later agent works on remote config. I think this inconsistent time can be optimized by cds sync task initial delay.

If you really care about the latency, why don't you inject the latest values through the system environment? The point of introducing dynamic configurations is to change the values in the runtime, rather than making all agents running by default and then updated later.

**Reply by weixiang1862** [reply]
Ok, thanks. Inject by system environment is a good way for this case.
